src/datagrid/DataGrid.tsx

```diff
@@ -759,9 +841,79 @@ export class DataGrid extends React.PureComponent<DataGridProps, DataGridState>
         return columns;
     }
 
+    // get details of first visible column
+    private getFirstVisibleColumn = () => {
+        const {allColumns} = this.state;
+        return allColumns.find((column: DataGridColumn) => column.isVisible === true);
+    };
+
+    // Check if datagrid need to render detail Pane for rows
+    private isDatagridWithDetailPane = (): boolean => {
+        const {rowType} = this.props;
+
+        return rowType
+            ? rowType === GridRowType.ROWS_WITH_DETAIL_PANE ||
+                  rowType === GridRowType.EXPANDABLE_ROWS_WITH_DETAIL_PANE ||
+                  rowType === GridRowType.COMPACT_ROWS_WITH_DETAIL_PANE
+            : false;
+    };
+
+    // Get details of row whose detail pane is open
+    private getRowDataWithOpenDetailPane = () => {
+        if (this.isDatagridWithDetailPane()) {
+            return this.state.allRows.find(
+                (row: DataGridRow) => row.detailPaneData && row.detailPaneData.isOpen === true,
+            );
+        }
+    };
+
+    // Check if any detail pane is open in datagrid
+    private isDetailPaneOpen = (): boolean => {
+        return this.isDatagridWithDetailPane() && this.getRowDataWithOpenDetailPane() !== undefined;
+    };
+
     /* ##########  DataGrid private methods end  ############ */
 
     /* ##########  DataGrid DOM methods start  ############ */
+
+    //funtion to render detail pane icon cell
+    private buildDetailPaneToggleIcon({rowID, detailPaneData}: DataGridRow): React.ReactElement {
+        const {id} = this.props;
+        const {isOpen, hideDetailPane} = detailPaneData
+            ? detailPaneData
+            : {
+                  isOpen: false,
+                  hideDetailPane: false,
+              };
+        return (
+            <div
+                className={classNames([
+                    ClassNames.DATAGRID_DETAIL_CARET, //prettier
+                    ClassNames.DATAGRID_FIXED_COLUMN,
+                    ClassNames.DATAGRID_CELL,
+                    ClassNames.DATAGRID_NG_STAR_INSERTED,
+                ])}
+            >
+                {!hideDetailPane && (
+                    <Button
+                        key={`${id}-${"detail-toggle"}-${rowID}`}
+                        defaultBtn={false}
+                        onClick={() => rowID !== undefined && this.handleDetailPaneToggle(rowID)}
+                        aria-haspopup="dialog"
+                        className={classNames([ClassNames.DATAGRID_DETAIL_CARET_BUTTON, isOpen && ClassNames.IS_OPEN])}
+                        aria-expanded={isOpen}
+                        aria-controls={`clr-id-${rowID}`}
+                        icon={{
+                            shape: "angle-double",
```

Comment:
can we create enum with these shapes ?  We have many such shapes I think in our codebase.

Reply:
We support quite lot of shapes. Can't add enum for all of them. Plus this seem unnecessary. Except some common icon like close we are using other shapes only one time